lw/co/generator.h

```diff
@@ -76,8 +85,123 @@ class Generator<T>::promise_type {
   }
 
 private:
-  T _current_value;
+  std::unique_ptr<T> _current_value;
   friend class Generator<T>;
 };
 
+template <typename T>
+class GeneratorIterator {
+public:
+  GeneratorIterator(Generator<T>* generator): _generator{generator} {}
+
+  GeneratorIterator& operator++() {
+    if (_generator && !_generator->next()) _generator = nullptr;
+    return *this;
+  }
+
+  T& operator*() { return _generator->value(); }
+  const T& operator*() const { return _generator->value(); }
+
+  template <typename U>
+  bool operator==(const GeneratorIterator<U>& other) {
+    return _generator == other._generator;
+  }
+
+private:
+  Generator<T>* _generator;
+};
+
+template <typename T>
+GeneratorIterator<T> Generator<T>::begin() {
+  if (!_handle.promise()._current_value) next();
+  return GeneratorIterator<T>{_handle.done() ? nullptr : this};
+}
+
+template <typename T>
+GeneratorIterator<T> Generator<T>::end() {
+  return GeneratorIterator<T>{nullptr};
+}
+
+// -------------------------------------------------------------------------- //
+
+template <typename T>
+class AsyncGenerator {
+public:
+  class promise_type;
+
+  AsyncGenerator(): _state{std::make_shared<SharedState>()} {};
+  ~AsyncGenerator() = default;
+
+  AsyncGenerator(AsyncGenerator&&) = default;
+  AsyncGenerator& operator=(AsyncGenerator&&) = default;
+
+  AsyncGenerator(const AsyncGenerator&) = delete;
+  AsyncGenerator& operator=(const AsyncGenerator&) = delete;
+
+  Future<bool> next() {
+    if (_state->done) return make_resolved_future(false);
+    _state->next = Promise<bool>{};
+    return _state->next.get_future();
+  }
+
+  const T& value() const {
+    return *_state->current_value;
+  }
+
+  T& value() {
+    return *_state->current_value;
+  }
+
+private:
+  struct SharedState {
```

Comment:
Why does your server need state?

Reply:
"State" is just another word for "data" or "information." The phrase "stateless server" is a bit misleading, because everything that handles information has state, even if only temporarily. The "state" for a stateless server is loaded, acted upon, and released in execution of the request coming in. It is only called "stateless" because the state for users and requests is not kept around in memory between requests.